{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from loguru import logger"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-04-29 16:09:59.966 | INFO     | pyMultiOmics.reactome:get_neo4j_driver:24 - Created graph database driver for bolt://localhost:7687 (neo4j)\n",
      "2021-04-29 16:10:01.318 | WARNING  | pyMultiOmics.pipelines:<module>:18 - Error loading rpy2 on win32, R function calls will be made online\n"
     ]
    }
   ],
   "source": [
    "sys.path.append('..')\n",
    "\n",
    "from pyMultiOmics.constants import *\n",
    "from pyMultiOmics.mapping import Mapper\n",
    "from pyMultiOmics.common import set_log_level_info\n",
    "from pyMultiOmics.analysis import *\n",
    "from pyMultiOmics.query import *\n",
    "from pyMultiOmics.pipelines import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Demonstration of pyMultiOmics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load the processed Zebrafish data from [1]\n",
    "\n",
    "[1] [Rabinowitz, Jeremy S., et al. \"Transcriptomic, proteomic, and metabolomic landscape of positional memory in the caudal fin of zebrafish.\" Proceedings of the National Academy of Sciences 114.5 (2017): E717-E726.](https://www.pnas.org/content/114/5/E717.short)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'C:\\\\Users\\\\joewa\\\\Work\\\\git\\\\pyMultiOmics\\\\notebooks\\\\test_data\\\\zebrafish_data'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATA_FOLDER = os.path.abspath(os.path.join('test_data', 'zebrafish_data'))\n",
    "DATA_FOLDER"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read transcriptomics data (identified by their gene ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_data = pd.read_csv(os.path.join(DATA_FOLDER, 'gene_data_combined.csv'), index_col='Identifier')\n",
    "gene_design = pd.read_csv(os.path.join(DATA_FOLDER, 'gene_design.csv'), index_col='sample')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read proteomics data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_data = pd.read_csv(os.path.join(DATA_FOLDER, 'protein_data.csv'), index_col='Uniprot')\n",
    "protein_design = pd.read_csv(os.path.join(DATA_FOLDER, 'protein_design.csv'), index_col='sample')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read metabolomics data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "compound_data = pd.read_csv(os.path.join(DATA_FOLDER, 'compound_data_chebi.csv'), index_col='Identifier')\n",
    "compound_design = pd.read_csv(os.path.join(DATA_FOLDER, 'compound_design.csv'), index_col='sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set_log_level_info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a mapping object"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The mapping object uses Reactome to map the different biological entities in the data:\n",
    "- Transcripts (or genes) are connected to the proteins they encode\n",
    "- Proteins and compounds are connected to reactions they're involved in\n",
    "- Reactions are connected to pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-04-29 16:10:02.161 | INFO     | pyMultiOmics.functions:remove_dupes:385 - Removing 0 rows with duplicate identifiers\n",
      "2021-04-29 16:10:02.161 | INFO     | pyMultiOmics.functions:reactome_mapping:78 - There are 124 observed compound ids\n",
      "2021-04-29 16:10:02.161 | INFO     | pyMultiOmics.functions:reactome_mapping:81 - Mapping genes -> proteins\n",
      "2021-04-29 16:10:03.025 | INFO     | pyMultiOmics.functions:reactome_mapping:86 - Mapping proteins -> reactions\n",
      "2021-04-29 16:10:04.654 | INFO     | pyMultiOmics.functions:reactome_mapping:94 - Mapping compounds -> reactions\n",
      "2021-04-29 16:10:05.522 | INFO     | pyMultiOmics.functions:reactome_mapping:100 - Mapping reactions -> pathways\n",
      "2021-04-29 16:10:05.831 | INFO     | pyMultiOmics.functions:reactome_mapping:111 - Mapping reactions -> proteins\n",
      "2021-04-29 16:10:07.926 | INFO     | pyMultiOmics.functions:reactome_mapping:118 - Mapping reactions -> compounds\n",
      "2021-04-29 16:10:09.025 | INFO     | pyMultiOmics.functions:reactome_mapping:130 - Mapping proteins -> genes\n",
      "2021-04-29 16:10:23.147 | INFO     | pyMultiOmics.mapping:_add_nodes:184 - Processing nodes: genes\n",
      "2021-04-29 16:10:25.398 | INFO     | pyMultiOmics.mapping:_add_nodes:184 - Processing nodes: proteins\n",
      "2021-04-29 16:10:26.208 | INFO     | pyMultiOmics.mapping:_add_nodes:184 - Processing nodes: compounds\n",
      "2021-04-29 16:10:26.350 | INFO     | pyMultiOmics.mapping:_add_nodes:184 - Processing nodes: reactions\n",
      "2021-04-29 16:10:26.837 | INFO     | pyMultiOmics.mapping:_add_nodes:184 - Processing nodes: pathways\n",
      "2021-04-29 16:10:26.857 | INFO     | pyMultiOmics.mapping:_add_edges:232 - Processing edges: gene_proteins\n",
      "2021-04-29 16:10:27.419 | INFO     | pyMultiOmics.mapping:_add_edges:232 - Processing edges: protein_reactions\n",
      "2021-04-29 16:10:30.785 | INFO     | pyMultiOmics.mapping:_add_edges:232 - Processing edges: compound_reactions\n",
      "2021-04-29 16:10:31.826 | INFO     | pyMultiOmics.mapping:_add_edges:232 - Processing edges: reaction_pathways\n",
      "2021-04-29 16:10:32.259 | INFO     | pyMultiOmics.mapping:build:84 - Created a multi-omics network with 20420 nodes and 21577 edges\n",
      "2021-04-29 16:10:32.417 | INFO     | pyMultiOmics.mapping:build:85 - node_counts = {'Genomics': 8956, 'Transcriptomics': 0, 'Proteomics': 7932, 'Metabolomics': 1695, 'Reactions': 1589, 'Pathways': 248}\n"
     ]
    }
   ],
   "source": [
    "m = Mapper(DANIO_RERIO, metabolic_pathway_only=True) \\\n",
    "        .set_gene(gene_data, gene_design) \\\n",
    "        .set_protein(protein_data, protein_design) \\\n",
    "        .set_compound(compound_data, compound_design) \\\n",
    "        .build()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DE analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "ap = AnalysisPipeline(m)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-04-29 16:10:32.704 | INFO     | pyMultiOmics.pipelines:run_ttest:134 - t-test case is Distal, control is Proximal\n",
      "2021-04-29 16:10:46.857 | INFO     | pyMultiOmics.pipelines:run_ttest:134 - t-test case is Distal, control is Middle\n",
      "2021-04-29 16:11:01.222 | INFO     | pyMultiOmics.pipelines:run_ttest:134 - t-test case is Proximal, control is Middle\n"
     ]
    }
   ],
   "source": [
    "method = INFERENCE_T_TEST\n",
    "ap.run_de(method, GENES, 'Distal', 'Proximal')\n",
    "ap.run_de(method, GENES, 'Distal', 'Middle')\n",
    "ap.run_de(method, GENES, 'Proximal', 'Middle')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-04-29 16:11:15.065 | INFO     | pyMultiOmics.pipelines:run_ttest:134 - t-test case is Distal, control is Proximal\n",
      "2021-04-29 16:11:16.668 | INFO     | pyMultiOmics.pipelines:run_ttest:134 - t-test case is Distal, control is Middle\n",
      "2021-04-29 16:11:18.165 | INFO     | pyMultiOmics.pipelines:run_ttest:134 - t-test case is Proximal, control is Middle\n",
      "2021-04-29 16:11:19.690 | INFO     | pyMultiOmics.pipelines:run_ttest:134 - t-test case is Distal, control is Proximal\n",
      "2021-04-29 16:11:19.784 | INFO     | pyMultiOmics.pipelines:run_ttest:134 - t-test case is Distal, control is Middle\n",
      "2021-04-29 16:11:19.846 | INFO     | pyMultiOmics.pipelines:run_ttest:134 - t-test case is Proximal, control is Middle\n"
     ]
    }
   ],
   "source": [
    "method = INFERENCE_T_TEST\n",
    "ap.run_de(method, PROTEINS, 'Distal', 'Proximal')\n",
    "ap.run_de(method, PROTEINS, 'Distal', 'Middle')\n",
    "ap.run_de(method, PROTEINS, 'Proximal', 'Middle')\n",
    "\n",
    "ap.run_de(method, COMPOUNDS, 'Distal', 'Proximal')\n",
    "ap.run_de(method, COMPOUNDS, 'Distal', 'Middle')\n",
    "ap.run_de(method, COMPOUNDS, 'Proximal', 'Middle')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Query results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Retrieve a single node"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>display_name</th>\n",
       "      <th>data_type</th>\n",
       "      <th>observed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>entity_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15366</th>\n",
       "      <td>Acetic acid</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          display_name  data_type  observed\n",
       "entity_id                                  \n",
       "15366      Acetic acid  compounds     False"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "node_id = '15366'\n",
    "res = QueryBuilder(ap) \\\n",
    "        .add(Entity(node_id)) \\\n",
    "        .run()\n",
    "res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Retrieve multiple nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>display_name</th>\n",
       "      <th>data_type</th>\n",
       "      <th>observed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>entity_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15366</th>\n",
       "      <td>Acetic acid</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSDARG00000037781</th>\n",
       "      <td>Acss2</td>\n",
       "      <td>genes</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F1QAA7</th>\n",
       "      <td>F1QAA7</td>\n",
       "      <td>proteins</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   display_name  data_type  observed\n",
       "entity_id                                           \n",
       "15366               Acetic acid  compounds     False\n",
       "ENSDARG00000037781        Acss2      genes      True\n",
       "F1QAA7                   F1QAA7   proteins      True"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "node_id = ['15366', 'ENSDARG00000037781', 'F1QAA7']\n",
    "res = QueryBuilder(ap) \\\n",
    "        .add(Entity(node_id)) \\\n",
    "        .run()\n",
    "res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Retrieve nodes connected to a single node"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>display_name</th>\n",
       "      <th>data_type</th>\n",
       "      <th>observed</th>\n",
       "      <th>source_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>entity_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSDARG00000037781</th>\n",
       "      <td>Acss2</td>\n",
       "      <td>genes</td>\n",
       "      <td>True</td>\n",
       "      <td>F1QAA7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15366</th>\n",
       "      <td>Acetic acid</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>F1QAA7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16027</th>\n",
       "      <td>Adenosine 5-monophosphate</td>\n",
       "      <td>compounds</td>\n",
       "      <td>True</td>\n",
       "      <td>F1QAA7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29888</th>\n",
       "      <td>Diphosphoric acid</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>F1QAA7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15351</th>\n",
       "      <td>Acetyl-coa</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>F1QAA7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15346</th>\n",
       "      <td>Coenzyme a</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>F1QAA7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30616</th>\n",
       "      <td>Atp(4-)</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>F1QAA7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-DRE-71735</th>\n",
       "      <td>Acetate + coa + atp =&gt; acetyl-coa + amp + pyro...</td>\n",
       "      <td>reactions</td>\n",
       "      <td>None</td>\n",
       "      <td>F1QAA7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-DRE-71384</th>\n",
       "      <td>Ethanol oxidation</td>\n",
       "      <td>pathways</td>\n",
       "      <td>None</td>\n",
       "      <td>F1QAA7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                         display_name  \\\n",
       "entity_id                                                               \n",
       "ENSDARG00000037781                                              Acss2   \n",
       "15366                                                     Acetic acid   \n",
       "16027                                       Adenosine 5-monophosphate   \n",
       "29888                                               Diphosphoric acid   \n",
       "15351                                                      Acetyl-coa   \n",
       "15346                                                      Coenzyme a   \n",
       "30616                                                         Atp(4-)   \n",
       "R-DRE-71735         Acetate + coa + atp => acetyl-coa + amp + pyro...   \n",
       "R-DRE-71384                                         Ethanol oxidation   \n",
       "\n",
       "                    data_type observed source_id  \n",
       "entity_id                                         \n",
       "ENSDARG00000037781      genes     True    F1QAA7  \n",
       "15366               compounds    False    F1QAA7  \n",
       "16027               compounds     True    F1QAA7  \n",
       "29888               compounds    False    F1QAA7  \n",
       "15351               compounds    False    F1QAA7  \n",
       "15346               compounds    False    F1QAA7  \n",
       "30616               compounds    False    F1QAA7  \n",
       "R-DRE-71735         reactions     None    F1QAA7  \n",
       "R-DRE-71384          pathways     None    F1QAA7  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query_id = 'F1QAA7'\n",
    "res = QueryBuilder(ap) \\\n",
    "        .add(Entity(query_id)) \\\n",
    "        .add(Connected()) \\\n",
    "        .run()\n",
    "res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Retrieve top-10 significantly changing genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "case = 'Distal'\n",
    "control = 'Proximal'\n",
    "pval = 0.05\n",
    "fc_lte = -2\n",
    "fc_gte = 2\n",
    "N = 20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>display_name</th>\n",
       "      <th>data_type</th>\n",
       "      <th>observed</th>\n",
       "      <th>padj_Distal_vs_Proximal</th>\n",
       "      <th>FC_Distal_vs_Proximal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSDARG00000045854</th>\n",
       "      <td>Fgf23</td>\n",
       "      <td>genes</td>\n",
       "      <td>True</td>\n",
       "      <td>3.569550e-10</td>\n",
       "      <td>24.000289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSDARG00000071091</th>\n",
       "      <td>Chrm3a</td>\n",
       "      <td>genes</td>\n",
       "      <td>True</td>\n",
       "      <td>3.307603e-11</td>\n",
       "      <td>22.486196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSDARG00000104259</th>\n",
       "      <td>Cabz01072885.1</td>\n",
       "      <td>genes</td>\n",
       "      <td>True</td>\n",
       "      <td>2.782933e-10</td>\n",
       "      <td>20.545646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSDARG00000005522</th>\n",
       "      <td>Galr1a</td>\n",
       "      <td>genes</td>\n",
       "      <td>True</td>\n",
       "      <td>2.091894e-09</td>\n",
       "      <td>20.531569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSDARG00000091254</th>\n",
       "      <td>Si:ch73-59p9.2</td>\n",
       "      <td>genes</td>\n",
       "      <td>True</td>\n",
       "      <td>3.736906e-10</td>\n",
       "      <td>19.689389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSDARG00000008541</th>\n",
       "      <td>Chia.4</td>\n",
       "      <td>genes</td>\n",
       "      <td>True</td>\n",
       "      <td>3.736906e-10</td>\n",
       "      <td>19.689389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSDARG00000093738</th>\n",
       "      <td>Pth1a</td>\n",
       "      <td>genes</td>\n",
       "      <td>True</td>\n",
       "      <td>3.736906e-10</td>\n",
       "      <td>19.689389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSDARG00000037946</th>\n",
       "      <td>Prl</td>\n",
       "      <td>genes</td>\n",
       "      <td>True</td>\n",
       "      <td>3.736906e-10</td>\n",
       "      <td>19.689389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSDARG00000036171</th>\n",
       "      <td>Rnasel3</td>\n",
       "      <td>genes</td>\n",
       "      <td>True</td>\n",
       "      <td>1.848146e-11</td>\n",
       "      <td>19.636782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSDARG00000026484</th>\n",
       "      <td>Rab15</td>\n",
       "      <td>genes</td>\n",
       "      <td>True</td>\n",
       "      <td>3.569550e-10</td>\n",
       "      <td>19.138412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSDARG00000102336</th>\n",
       "      <td>Slc5a7</td>\n",
       "      <td>genes</td>\n",
       "      <td>True</td>\n",
       "      <td>5.646605e-09</td>\n",
       "      <td>-21.542885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSDARG00000070038</th>\n",
       "      <td>Rbp2a</td>\n",
       "      <td>genes</td>\n",
       "      <td>True</td>\n",
       "      <td>1.821398e-10</td>\n",
       "      <td>-21.882546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSDARG00000067859</th>\n",
       "      <td>Scospondin</td>\n",
       "      <td>genes</td>\n",
       "      <td>True</td>\n",
       "      <td>1.625297e-10</td>\n",
       "      <td>-22.444017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSDARG00000008306</th>\n",
       "      <td>Rdh5</td>\n",
       "      <td>genes</td>\n",
       "      <td>True</td>\n",
       "      <td>7.680307e-11</td>\n",
       "      <td>-22.449311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSDARG00000100258</th>\n",
       "      <td>Si:dkeyp-104b3.2</td>\n",
       "      <td>genes</td>\n",
       "      <td>True</td>\n",
       "      <td>2.760891e-09</td>\n",
       "      <td>-22.482546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSDARG00000093089</th>\n",
       "      <td>Tac3a</td>\n",
       "      <td>genes</td>\n",
       "      <td>True</td>\n",
       "      <td>3.569550e-10</td>\n",
       "      <td>-22.875241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSDARG00000038894</th>\n",
       "      <td>Tmx3a</td>\n",
       "      <td>genes</td>\n",
       "      <td>True</td>\n",
       "      <td>4.745826e-10</td>\n",
       "      <td>-23.095898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSDARG00000039067</th>\n",
       "      <td>Srd5a2b</td>\n",
       "      <td>genes</td>\n",
       "      <td>True</td>\n",
       "      <td>6.493330e-11</td>\n",
       "      <td>-23.483296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSDARG00000076270</th>\n",
       "      <td>Adamts13</td>\n",
       "      <td>genes</td>\n",
       "      <td>True</td>\n",
       "      <td>6.923704e-11</td>\n",
       "      <td>-23.815791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSDARG00000025783</th>\n",
       "      <td>Si:ch211-125e6.11</td>\n",
       "      <td>genes</td>\n",
       "      <td>True</td>\n",
       "      <td>6.804488e-11</td>\n",
       "      <td>-26.987938</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         display_name data_type  observed  \\\n",
       "ENSDARG00000045854              Fgf23     genes      True   \n",
       "ENSDARG00000071091             Chrm3a     genes      True   \n",
       "ENSDARG00000104259     Cabz01072885.1     genes      True   \n",
       "ENSDARG00000005522             Galr1a     genes      True   \n",
       "ENSDARG00000091254     Si:ch73-59p9.2     genes      True   \n",
       "ENSDARG00000008541             Chia.4     genes      True   \n",
       "ENSDARG00000093738              Pth1a     genes      True   \n",
       "ENSDARG00000037946                Prl     genes      True   \n",
       "ENSDARG00000036171            Rnasel3     genes      True   \n",
       "ENSDARG00000026484              Rab15     genes      True   \n",
       "ENSDARG00000102336             Slc5a7     genes      True   \n",
       "ENSDARG00000070038              Rbp2a     genes      True   \n",
       "ENSDARG00000067859         Scospondin     genes      True   \n",
       "ENSDARG00000008306               Rdh5     genes      True   \n",
       "ENSDARG00000100258   Si:dkeyp-104b3.2     genes      True   \n",
       "ENSDARG00000093089              Tac3a     genes      True   \n",
       "ENSDARG00000038894              Tmx3a     genes      True   \n",
       "ENSDARG00000039067            Srd5a2b     genes      True   \n",
       "ENSDARG00000076270           Adamts13     genes      True   \n",
       "ENSDARG00000025783  Si:ch211-125e6.11     genes      True   \n",
       "\n",
       "                    padj_Distal_vs_Proximal  FC_Distal_vs_Proximal  \n",
       "ENSDARG00000045854             3.569550e-10              24.000289  \n",
       "ENSDARG00000071091             3.307603e-11              22.486196  \n",
       "ENSDARG00000104259             2.782933e-10              20.545646  \n",
       "ENSDARG00000005522             2.091894e-09              20.531569  \n",
       "ENSDARG00000091254             3.736906e-10              19.689389  \n",
       "ENSDARG00000008541             3.736906e-10              19.689389  \n",
       "ENSDARG00000093738             3.736906e-10              19.689389  \n",
       "ENSDARG00000037946             3.736906e-10              19.689389  \n",
       "ENSDARG00000036171             1.848146e-11              19.636782  \n",
       "ENSDARG00000026484             3.569550e-10              19.138412  \n",
       "ENSDARG00000102336             5.646605e-09             -21.542885  \n",
       "ENSDARG00000070038             1.821398e-10             -21.882546  \n",
       "ENSDARG00000067859             1.625297e-10             -22.444017  \n",
       "ENSDARG00000008306             7.680307e-11             -22.449311  \n",
       "ENSDARG00000100258             2.760891e-09             -22.482546  \n",
       "ENSDARG00000093089             3.569550e-10             -22.875241  \n",
       "ENSDARG00000038894             4.745826e-10             -23.095898  \n",
       "ENSDARG00000039067             6.493330e-11             -23.483296  \n",
       "ENSDARG00000076270             6.923704e-11             -23.815791  \n",
       "ENSDARG00000025783             6.804488e-11             -26.987938  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res = QueryBuilder(ap) \\\n",
    "        .add(Select(GENES)) \\\n",
    "        .add(SignificantDE(case, control, pval, fc_lte=fc_lte, fc_gte=fc_gte, N=N)) \\\n",
    "        .run()\n",
    "res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find the compounds that are connected to the DE genes above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>display_name</th>\n",
       "      <th>data_type</th>\n",
       "      <th>observed</th>\n",
       "      <th>source_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>entity_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>78579</th>\n",
       "      <td>Aflatoxin b1 endo-cysteinylglycine conjugate</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000091254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29985</th>\n",
       "      <td>L-glutamate(1-)</td>\n",
       "      <td>compounds</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSDARG00000091254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15377</th>\n",
       "      <td>Water</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000091254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78584</th>\n",
       "      <td>Aflatoxin b1 exo-cysteinylglycine conjugate</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000091254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78587</th>\n",
       "      <td>Aflatoxin b1 exo glutathione conjugate</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000091254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78581</th>\n",
       "      <td>Aflatoxin b1 endo glutathione conjugate</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000091254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16978</th>\n",
       "      <td>Leukotriene c4</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000091254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28666</th>\n",
       "      <td>Leukotriene d4</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000091254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4047</th>\n",
       "      <td>L-cysteinylglycine</td>\n",
       "      <td>compounds</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSDARG00000091254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16856</th>\n",
       "      <td>Glutathione</td>\n",
       "      <td>compounds</td>\n",
       "      <td>True</td>\n",
       "      <td>ENSDARG00000091254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17898</th>\n",
       "      <td>All-trans-retinal</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000070038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16474</th>\n",
       "      <td>Nadph</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000070038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18009</th>\n",
       "      <td>Nadp(+)</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000070038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17336</th>\n",
       "      <td>All-trans-retinol</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000070038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15378</th>\n",
       "      <td>Hydron</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000070038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70762</th>\n",
       "      <td>All-trans-retinyl linoleate</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000070038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45021</th>\n",
       "      <td>Palmitoyl group</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000070038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25667</th>\n",
       "      <td>Oleoyl group</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000070038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70761</th>\n",
       "      <td>All-trans-retinyl stearate</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000070038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26753</th>\n",
       "      <td>Stearoyl group</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000070038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17616</th>\n",
       "      <td>All-trans-retinyl palmitate</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000070038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32386</th>\n",
       "      <td>Linoleoyl group</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000070038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70760</th>\n",
       "      <td>All-trans-retinyl oleate</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000070038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16330</th>\n",
       "      <td>17beta-hydroxy-5alpha-androstan-3-one</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000039067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18009</th>\n",
       "      <td>Nadp(+)</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000039067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17347</th>\n",
       "      <td>Testosterone</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000039067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15378</th>\n",
       "      <td>Hydron</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000039067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16474</th>\n",
       "      <td>Nadph</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000039067</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           display_name  data_type  observed  \\\n",
       "entity_id                                                                      \n",
       "78579      Aflatoxin b1 endo-cysteinylglycine conjugate  compounds     False   \n",
       "29985                                   L-glutamate(1-)  compounds      True   \n",
       "15377                                             Water  compounds     False   \n",
       "78584       Aflatoxin b1 exo-cysteinylglycine conjugate  compounds     False   \n",
       "78587            Aflatoxin b1 exo glutathione conjugate  compounds     False   \n",
       "78581           Aflatoxin b1 endo glutathione conjugate  compounds     False   \n",
       "16978                                    Leukotriene c4  compounds     False   \n",
       "28666                                    Leukotriene d4  compounds     False   \n",
       "4047                                 L-cysteinylglycine  compounds      True   \n",
       "16856                                       Glutathione  compounds      True   \n",
       "17898                                 All-trans-retinal  compounds     False   \n",
       "16474                                             Nadph  compounds     False   \n",
       "18009                                           Nadp(+)  compounds     False   \n",
       "17336                                 All-trans-retinol  compounds     False   \n",
       "15378                                            Hydron  compounds     False   \n",
       "70762                       All-trans-retinyl linoleate  compounds     False   \n",
       "45021                                   Palmitoyl group  compounds     False   \n",
       "25667                                      Oleoyl group  compounds     False   \n",
       "70761                        All-trans-retinyl stearate  compounds     False   \n",
       "26753                                    Stearoyl group  compounds     False   \n",
       "17616                       All-trans-retinyl palmitate  compounds     False   \n",
       "32386                                   Linoleoyl group  compounds     False   \n",
       "70760                          All-trans-retinyl oleate  compounds     False   \n",
       "16330             17beta-hydroxy-5alpha-androstan-3-one  compounds     False   \n",
       "18009                                           Nadp(+)  compounds     False   \n",
       "17347                                      Testosterone  compounds     False   \n",
       "15378                                            Hydron  compounds     False   \n",
       "16474                                             Nadph  compounds     False   \n",
       "\n",
       "                    source_id  \n",
       "entity_id                      \n",
       "78579      ENSDARG00000091254  \n",
       "29985      ENSDARG00000091254  \n",
       "15377      ENSDARG00000091254  \n",
       "78584      ENSDARG00000091254  \n",
       "78587      ENSDARG00000091254  \n",
       "78581      ENSDARG00000091254  \n",
       "16978      ENSDARG00000091254  \n",
       "28666      ENSDARG00000091254  \n",
       "4047       ENSDARG00000091254  \n",
       "16856      ENSDARG00000091254  \n",
       "17898      ENSDARG00000070038  \n",
       "16474      ENSDARG00000070038  \n",
       "18009      ENSDARG00000070038  \n",
       "17336      ENSDARG00000070038  \n",
       "15378      ENSDARG00000070038  \n",
       "70762      ENSDARG00000070038  \n",
       "45021      ENSDARG00000070038  \n",
       "25667      ENSDARG00000070038  \n",
       "70761      ENSDARG00000070038  \n",
       "26753      ENSDARG00000070038  \n",
       "17616      ENSDARG00000070038  \n",
       "32386      ENSDARG00000070038  \n",
       "70760      ENSDARG00000070038  \n",
       "16330      ENSDARG00000039067  \n",
       "18009      ENSDARG00000039067  \n",
       "17347      ENSDARG00000039067  \n",
       "15378      ENSDARG00000039067  \n",
       "16474      ENSDARG00000039067  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res = QueryBuilder(ap) \\\n",
    "        .add(Select(GENES)) \\\n",
    "        .add(SignificantDE(case, control, pval, fc_lte=fc_lte, fc_gte=fc_gte, N=N)) \\\n",
    "        .add(Connected(data_type=COMPOUNDS)) \\\n",
    "        .run()\n",
    "res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Retrieve entity info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>display_name</th>\n",
       "      <th>data_type</th>\n",
       "      <th>observed</th>\n",
       "      <th>source_id</th>\n",
       "      <th>infos</th>\n",
       "      <th>images</th>\n",
       "      <th>links</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>entity_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Q1L8M3</th>\n",
       "      <td>Q1L8M3</td>\n",
       "      <td>proteins</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000045854</td>\n",
       "      <td>[{'key': 'Name', 'value': 'Q1L8M3'}, {'key': '...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[{'text': 'Link to UniProt', 'href': 'http://w...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X1WHZ7</th>\n",
       "      <td>X1WHZ7</td>\n",
       "      <td>proteins</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000071091</td>\n",
       "      <td>[{'key': 'Name', 'value': 'Muscarinic acetylch...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[{'text': 'Link to UniProt', 'href': 'http://w...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0A0G2L3I9</th>\n",
       "      <td>A0A0G2L3I9</td>\n",
       "      <td>proteins</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000104259</td>\n",
       "      <td>[{'key': 'Name', 'value': 'A0A0G2L3I9'}, {'key...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[{'text': 'Link to UniProt', 'href': 'http://w...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F1QSA8</th>\n",
       "      <td>F1QSA8</td>\n",
       "      <td>proteins</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000005522</td>\n",
       "      <td>[{'key': 'Name', 'value': 'F1QSA8'}, {'key': '...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[{'text': 'Link to UniProt', 'href': 'http://w...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E7F2C6</th>\n",
       "      <td>E7F2C6</td>\n",
       "      <td>proteins</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000091254</td>\n",
       "      <td>[{'key': 'Name', 'value': 'E7F2C6'}, {'key': '...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[{'text': 'Link to UniProt', 'href': 'http://w...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16474</th>\n",
       "      <td>Nadph</td>\n",
       "      <td>compounds</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000039067</td>\n",
       "      <td>[{'key': 'PiMP Peak ID', 'value': 'None'}, {'k...</td>\n",
       "      <td>[http://www.ebi.ac.uk/chebi/displayImage.do?de...</td>\n",
       "      <td>[{'text': 'Link to ChEBI database', 'href': 'h...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-DRE-469659</th>\n",
       "      <td>Testosterone is converted to 5-alpha-dihydroxy...</td>\n",
       "      <td>reactions</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSDARG00000039067</td>\n",
       "      <td>[{'key': 'Summary', 'value': 'The conversion o...</td>\n",
       "      <td>[https://reactome.org/ContentService/exporter/...</td>\n",
       "      <td>[{'text': 'Link to Reactome database', 'href':...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R-DRE-193048</th>\n",
       "      <td>Androgen biosynthesis</td>\n",
       "      <td>pathways</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSDARG00000039067</td>\n",
       "      <td>[{'key': 'Summary', 'value': 'Androgens are th...</td>\n",
       "      <td>[https://reactome.org/ContentService/exporter/...</td>\n",
       "      <td>[{'text': 'Link to Reactome database', 'href':...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0A286Y999</th>\n",
       "      <td>A0A286Y999</td>\n",
       "      <td>proteins</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000076270</td>\n",
       "      <td>[{'key': 'Name', 'value': 'A0A286Y999'}, {'key...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[{'text': 'Link to UniProt', 'href': 'http://w...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B3DIG2</th>\n",
       "      <td>B3DIG2</td>\n",
       "      <td>proteins</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSDARG00000025783</td>\n",
       "      <td>[{'key': 'Name', 'value': 'B3DIG2'}, {'key': '...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[{'text': 'Link to UniProt', 'href': 'http://w...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   display_name  data_type  \\\n",
       "entity_id                                                                    \n",
       "Q1L8M3                                                   Q1L8M3   proteins   \n",
       "X1WHZ7                                                   X1WHZ7   proteins   \n",
       "A0A0G2L3I9                                           A0A0G2L3I9   proteins   \n",
       "F1QSA8                                                   F1QSA8   proteins   \n",
       "E7F2C6                                                   E7F2C6   proteins   \n",
       "...                                                         ...        ...   \n",
       "16474                                                     Nadph  compounds   \n",
       "R-DRE-469659  Testosterone is converted to 5-alpha-dihydroxy...  reactions   \n",
       "R-DRE-193048                              Androgen biosynthesis   pathways   \n",
       "A0A286Y999                                           A0A286Y999   proteins   \n",
       "B3DIG2                                                   B3DIG2   proteins   \n",
       "\n",
       "             observed           source_id  \\\n",
       "entity_id                                   \n",
       "Q1L8M3          False  ENSDARG00000045854   \n",
       "X1WHZ7          False  ENSDARG00000071091   \n",
       "A0A0G2L3I9      False  ENSDARG00000104259   \n",
       "F1QSA8          False  ENSDARG00000005522   \n",
       "E7F2C6          False  ENSDARG00000091254   \n",
       "...               ...                 ...   \n",
       "16474           False  ENSDARG00000039067   \n",
       "R-DRE-469659     None  ENSDARG00000039067   \n",
       "R-DRE-193048     None  ENSDARG00000039067   \n",
       "A0A286Y999      False  ENSDARG00000076270   \n",
       "B3DIG2          False  ENSDARG00000025783   \n",
       "\n",
       "                                                          infos  \\\n",
       "entity_id                                                         \n",
       "Q1L8M3        [{'key': 'Name', 'value': 'Q1L8M3'}, {'key': '...   \n",
       "X1WHZ7        [{'key': 'Name', 'value': 'Muscarinic acetylch...   \n",
       "A0A0G2L3I9    [{'key': 'Name', 'value': 'A0A0G2L3I9'}, {'key...   \n",
       "F1QSA8        [{'key': 'Name', 'value': 'F1QSA8'}, {'key': '...   \n",
       "E7F2C6        [{'key': 'Name', 'value': 'E7F2C6'}, {'key': '...   \n",
       "...                                                         ...   \n",
       "16474         [{'key': 'PiMP Peak ID', 'value': 'None'}, {'k...   \n",
       "R-DRE-469659  [{'key': 'Summary', 'value': 'The conversion o...   \n",
       "R-DRE-193048  [{'key': 'Summary', 'value': 'Androgens are th...   \n",
       "A0A286Y999    [{'key': 'Name', 'value': 'A0A286Y999'}, {'key...   \n",
       "B3DIG2        [{'key': 'Name', 'value': 'B3DIG2'}, {'key': '...   \n",
       "\n",
       "                                                         images  \\\n",
       "entity_id                                                         \n",
       "Q1L8M3                                                       []   \n",
       "X1WHZ7                                                       []   \n",
       "A0A0G2L3I9                                                   []   \n",
       "F1QSA8                                                       []   \n",
       "E7F2C6                                                       []   \n",
       "...                                                         ...   \n",
       "16474         [http://www.ebi.ac.uk/chebi/displayImage.do?de...   \n",
       "R-DRE-469659  [https://reactome.org/ContentService/exporter/...   \n",
       "R-DRE-193048  [https://reactome.org/ContentService/exporter/...   \n",
       "A0A286Y999                                                   []   \n",
       "B3DIG2                                                       []   \n",
       "\n",
       "                                                          links  \n",
       "entity_id                                                        \n",
       "Q1L8M3        [{'text': 'Link to UniProt', 'href': 'http://w...  \n",
       "X1WHZ7        [{'text': 'Link to UniProt', 'href': 'http://w...  \n",
       "A0A0G2L3I9    [{'text': 'Link to UniProt', 'href': 'http://w...  \n",
       "F1QSA8        [{'text': 'Link to UniProt', 'href': 'http://w...  \n",
       "E7F2C6        [{'text': 'Link to UniProt', 'href': 'http://w...  \n",
       "...                                                         ...  \n",
       "16474         [{'text': 'Link to ChEBI database', 'href': 'h...  \n",
       "R-DRE-469659  [{'text': 'Link to Reactome database', 'href':...  \n",
       "R-DRE-193048  [{'text': 'Link to Reactome database', 'href':...  \n",
       "A0A286Y999    [{'text': 'Link to UniProt', 'href': 'http://w...  \n",
       "B3DIG2        [{'text': 'Link to UniProt', 'href': 'http://w...  \n",
       "\n",
       "[62 rows x 7 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res = QueryBuilder(ap) \\\n",
    "        .add(Select(GENES)) \\\n",
    "        .add(SignificantDE(case, control, pval, fc_lte=fc_lte, fc_gte=fc_gte, N=N)) \\\n",
    "        .add(Connected()) \\\n",
    "        .add(Info()) \\\n",
    "        .run()\n",
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyMultiOmics.info import get_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSDARG00000091254 genes\n",
      "{'infos': [{'key': 'Description', 'value': 'si:ch73-59p9.2 '}, {'key': 'Species', 'value': 'danio_rerio'}], 'links': [{'text': 'Link to Ensembl', 'href': 'https://www.ensembl.org/id/ENSDARG00000091254'}, {'text': 'Link to GeneCard', 'href': 'https://www.genecards.org/cgi-bin/carddisp.pl?gene=si:ch73-59p9.2'}, {'text': 'Transcript: si:ch73-59p9.2-201', 'href': 'https://www.ensembl.org/id/ENSDART00000111526'}], 'images': []}\n",
      "\n",
      "F1QAA7 proteins\n",
      "{'infos': [{'key': 'Name', 'value': 'Propionate--CoA ligase'}, {'key': 'EC Number', 'value': 'EC6.2.1.17'}, {'key': 'Catalytic Activity', 'value': '\\n\\nATP + CoA + propanoate = AMP + diphosphate + propanoyl-CoA\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n'}, {'key': 'Catalytic Activity', 'value': '\\n\\nacetate + ATP + CoA = acetyl-CoA + AMP + diphosphate\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n\\n'}, {'key': 'Gene_ontologies', 'value': 'acetate-CoA ligase activity; propionate-CoA ligase activity'}], 'links': [{'text': 'Link to UniProt', 'href': 'http://www.uniprot.org/uniprot/F1QAA7'}, {'text': 'Link to SWISS-MODEL', 'href': 'https://swissmodel.expasy.org/repository/uniprot/F1QAA7'}], 'images': []}\n",
      "\n",
      "15378 compounds\n",
      "{'infos': [{'key': 'PiMP Peak ID', 'value': 'None'}, {'key': 'KEGG ID', 'value': C00080}, {'key': 'FORMULA', 'value': H}, {'key': 'ChEBI ID', 'value': CHEBI:15378}, {'key': 'Definition', 'value': The general name for the hydrogen nucleus, to be used without regard to the hydrogen nuclear mass (either for hydrogen in its natural abundance or where it is not desired to distinguish between the isotopes).}, {'key': 'Monoisotopic Mass', 'value': 1.008}, {'key': 'SMILES', 'value': [H+]}, {'key': 'Inchi', 'value': InChI=1S/p+1}, {'key': 'InchiKey', 'value': ''}], 'images': ['http://www.ebi.ac.uk/chebi/displayImage.do?defaultImage=true&imageIndex=0&chebiId=15378'], 'links': [{'text': 'Link to ChEBI database', 'href': 'https://www.ebi.ac.uk/chebi/searchId.do?chebiId=CHEBI:15378'}, {'text': 'Link to KEGG COMPOUND database', 'href': 'http://www.genome.jp/dbget-bin/www_bget?cpd:C00080'}]}\n",
      "\n",
      "R-DRE-469659 reactions\n",
      "{'infos': [{'key': 'Summary', 'value': 'The conversion of testosterone to the most potent androgen, 5-alpha-dihydrotestosterone (DHT), is catalyzed by the microsomal 5alpha-steroid reductase enzymes, of which there are three reported types in humans to date (SRD5A1-3) (Andersson S and Russell DW, 1990, Andersson S et al, 1991, Uemura M et al, 2008 respectively). These enzymes are expressed in the prostate and other androgen target sites. Defects in SRD5A2 are the cause of pseudovaginal perineoscrotal hypospadias, also known as male pseudohermaphroditism (Anwar R et al, 1997). Corticotropin (Adrenocorticotropic hormone, ACTH) acts through the ACTH receptor called melanocortin receptor type 2 (MC2R) to stimulate steroidogenesis, increasing the production of androgens (McKenna et al, 1997).'}, {'key': 'Species', 'value': 'Danio rerio'}, {'key': 'Inferred', 'value': 'Inferred from Homo sapiens'}, {'key': 'Catalystactivity', 'value': '3-oxo-5-alpha-steroid 4-dehydrogenase activity of SRD5A1-3 [endoplasmic reticulum membrane]', 'url': ''}, {'key': 'Input', 'value': 'TEST [cytosol];NADPH [cytosol];H+ [cytosol]', 'url': 'https://reactome.org/content/detail/R-ALL-193057;https://reactome.org/content/detail/R-ALL-29364;https://reactome.org/content/detail/R-ALL-70106'}, {'key': 'Output', 'value': 'NADP+ [cytosol];DHTEST [cytosol]', 'url': 'https://reactome.org/content/detail/R-ALL-29366;https://reactome.org/content/detail/R-ALL-469662'}, {'key': 'Regulatedby', 'value': \"Positive regulation by 'POMC(138-176) [cytosol]'\", 'url': ''}], 'images': ['https://reactome.org/ContentService/exporter/diagram/R-DRE-469659.jpg?sel=R-DRE-469659&quality=7'], 'links': [{'text': 'Link to Reactome database', 'href': 'https://reactome.org/content/detail/R-DRE-469659'}]}\n",
      "\n",
      "R-DRE-174403 pathways\n",
      "{'infos': [{'key': 'Summary', 'value': 'The combination of glutamate, cysteine and ATP is required to form glutathione. The steps involved in the synthesis and recycling of glutathione are outlined (Meister, 1988).'}, {'key': 'Species', 'value': 'Danio rerio'}, {'key': 'Inferred', 'value': 'Inferred from Homo sapiens'}, {'key': 'Reactions', 'value': 'CHAC1,2 cleaves GSH to OPRO and CysGly;CNDP2:2Mn2+ dimer hydrolyses CysGly;GCL ligates L-Glu to L-Cys;GGCT transforms gGluCys to OPRO;GGT dimers hydrolyse GSH;GSS:Mg2+ dimer synthesizes GSH;OPLAH hydrolyses OPRO to L-Glu', 'url': 'https://reactome.org/content/detail/R-DRE-1247910;https://reactome.org/content/detail/R-DRE-1247922;https://reactome.org/content/detail/R-DRE-1247935;https://reactome.org/content/detail/R-DRE-174367;https://reactome.org/content/detail/R-DRE-174394;https://reactome.org/content/detail/R-DRE-6785928;https://reactome.org/content/detail/R-DRE-8943279'}], 'images': ['https://reactome.org/ContentService/exporter/diagram/R-DRE-174403.jpg?sel=R-DRE-174403'], 'links': [{'text': 'Link to Reactome database', 'href': 'https://reactome.org/content/detail/R-DRE-174403'}, {'text': 'SBML Export', 'href': 'https://reactome.org/ContentService/exporter/sbml/R-DRE-174403.xml'}]}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "entity_ids = ['ENSDARG00000091254', 'F1QAA7', '15378', 'R-DRE-469659', 'R-DRE-174403']\n",
    "data_types = ['genes', 'proteins', 'compounds', 'reactions', 'pathways']\n",
    "for entity_id, data_type in zip(entity_ids, data_types):\n",
    "    print(entity_id, data_type)\n",
    "    print(get_info(entity_id, data_type))\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot some heatmap using Plotly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res = QueryBuilder(ap) \\\n",
    "        .add(Select(GENES)) \\\n",
    "        .add(SignificantDE(case, control, pval, fc_lte=fc_lte, fc_gte=fc_gte, N=N)) \\\n",
    "        .run()\n",
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_type = GENES\n",
    "analysis = ap.get_de_analysis(data_type, case, control)\n",
    "wi = analysis.wi\n",
    "data_df, design_df = wi.data_df, wi.design_df\n",
    "data_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "case_group = design_df[design_df['group'] == case].index.values.tolist()\n",
    "control_group = design_df[design_df['group'] == control].index.values.tolist()\n",
    "selection = case_group + control_group\n",
    "selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "heatmap_df = wi.data_df.loc[res.index.values]\n",
    "heatmap_df = heatmap_df[selection]\n",
    "heatmap_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from plotly import express as px\n",
    "px.imshow(wi.standardize_df(heatmap_df, log=True, axis=0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
